File: app/src/main/java/android/freelessons/org/sampleandroidappusingfirebase/ui/EventActivity.java
package android.freelessons.org.sampleandroidappusingfirebase.ui;

import android.app.DatePickerDialog;
import android.freelessons.org.sampleandroidappusingfirebase.R;
import android.freelessons.org.sampleandroidappusingfirebase.domain.Event;
import android.freelessons.org.sampleandroidappusingfirebase.session.SessionManager;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.gms.common.api.Status;
import com.google.android.gms.location.places.Place;
import com.google.android.gms.location.places.ui.PlaceAutocompleteFragment;
import com.google.android.gms.location.places.ui.PlaceSelectionListener;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Locale;



public class EventActivity extends AppCompatActivity {
    private static String TAG="SAPWF";
    DatabaseReference databaseReference;
    EditText nameEditText;
    EditText descriptionEditText;
    Button saveEvent;
    EditText locationEditText;
    EditText startDateEditText;
    PlaceAutocompleteFragment locationView;


    SimpleDateFormat dateFormat = new SimpleDateFormat("dd MMM, yyyy", Locale.getDefault());
    Calendar myCalendar = Calendar.getInstance();
    Event event = new Event();
    SessionManager sessionManager;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.event);
        startDateEditText = findViewById(R.id.startDateEditText);
        locationEditText = findViewById(R.id.locationEditText);
        saveEvent = findViewById(R.id.saveEvent);
        descriptionEditText = findViewById(R.id.descriptionEditText);
        nameEditText = findViewById(R.id.nameEditText);
        sessionManager = new SessionManager(this);
        event = sessionManager.getEvent();
        myCalendar.setTime(event.getStartDate());
        databaseReference= FirebaseDatabase.getInstance().getReference();
        ActionBar actionBar = getSupportActionBar();
        if(actionBar!=null){
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
        locationView = (PlaceAutocompleteFragment) getFragmentManager().findFragmentById(R.id.locationView);
        locationView.setText("Search Location");
        locationView.setOnPlaceSelectedListener(new PlaceSelectionListener() {
            @Override
            public void onPlaceSelected(Place place) {
                Log.d(TAG,"Place:"+place.getName());
                Log.d(TAG,"Latitude:"+place.getLatLng().latitude);
                Log.d(TAG,"Longitude:"+place.getLatLng().longitude);
                locationEditText.setText(place.getName());
            }

            @Override
            public void onError(Status status) {
                Log.e(TAG,"Error: "+status);
            }

        });

    }

    @Override
    protected void onResume() {
        super.onResume();

    }

    @Override
    protected void onStart() {
        super.onStart();
        updateViews();
    }

    //update views
    private void updateViews() {
        nameEditText.setText(event.getName());
        descriptionEditText.setText(event.getDescription());
        locationEditText.setText(event.getLocation());
        startDateEditText.setText(dateFormat.format(event.getStartDate()));
        saveEvent.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(isValid()) {
                    saveEvent();
                }else{
                    Toast.makeText(getApplicationContext(),"Please make sure all fields are filled",Toast.LENGTH_LONG).show();
                }
            }
        });
        final DatePickerDialog.OnDateSetListener date = new DatePickerDialog.OnDateSetListener() {

            @Override
            public void onDateSet(DatePicker view, int year, int monthOfYear,
                                  int dayOfMonth) {
                myCalendar.set(Calendar.YEAR, year);
                myCalendar.set(Calendar.MONTH, monthOfYear);
                myCalendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);
                updateLabel();
            }

        };
        startDateEditText.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if(hasFocus){
                    new DatePickerDialog(EventActivity.this, date, myCalendar
                            .get(Calendar.YEAR), myCalendar.get(Calendar.MONTH),
                            myCalendar.get(Calendar.DAY_OF_MONTH)).show();
                }
            }
        });



    }
    private boolean isValid(){
        return !nameEditText.getText().toString().isEmpty() && !descriptionEditText.getText().toString().isEmpty();
    }
    private void saveEvent(){
        DatabaseReference eventReference;
        if(event.getEventId()==null) {
            eventReference = databaseReference.child("events").push();
            event.setEventId(eventReference.getKey());
        }else{
            eventReference=databaseReference.child("events").child(event.getEventId());
        }
        event.setName(nameEditText.getText().toString());
        event.setDescription(descriptionEditText.getText().toString());
        event.setLocation(locationEditText.getText().toString());
        if(FirebaseAuth.getInstance().getCurrentUser()!=null) {
            FirebaseUser firebaseUser = FirebaseAuth.getInstance().getCurrentUser();
            if (firebaseUser!=null) {
                event.setUserId(firebaseUser.getUid());
            }
        }
        try {
            event.setStartDate(dateFormat.parse(startDateEditText.getText().toString()));
        } catch (ParseException e) {
            //TODO: do something serious
            //e.printStackTrace();
        }

        eventReference.updateChildren(event.toMap()).addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if(task.isSuccessful()){
                    Log.d(TAG,"Success");
                    Toast.makeText(EventActivity.this, "Event saved successfully", Toast.LENGTH_SHORT).show();
                }else{
                    Log.e(TAG,"erro");
                    String errorMessage = null;
                    Exception exception = task.getException();
                    if (exception!=null) {
                        errorMessage = exception.getMessage();
                    }
                    Toast.makeText(EventActivity.this, "Error saving Event "+errorMessage, Toast.LENGTH_SHORT).show();
                }
            }
        });
    }
    private void updateLabel() {

        startDateEditText.setText(dateFormat.format(myCalendar.getTime()));
    }
}

File: app/src/main/java/android/freelessons/org/sampleandroidappusingfirebase/ui/EventActivity.kt
package android.freelessons.org.sampleandroidappusingfirebase.ui

import android.app.DatePickerDialog
import android.app.DatePickerDialog.OnDateSetListener
import android.freelessons.org.sampleandroidappusingfirebase.R
import android.freelessons.org.sampleandroidappusingfirebase.domain.Event
import android.freelessons.org.sampleandroidappusingfirebase.session.SessionManager
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.View.OnFocusChangeListener
import android.widget.Button
import android.widget.DatePicker
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.android.gms.common.api.Status
import com.google.android.gms.tasks.Task
import com.google.android.libraries.places.api.Places
import com.google.android.libraries.places.api.model.Place
import com.google.android.libraries.places.widget.AutocompleteSupportFragment
import com.google.android.libraries.places.widget.listener.PlaceSelectionListener
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

class EventActivity : AppCompatActivity() {
    private var databaseReference: DatabaseReference? = null
    private var nameEditText: EditText? = null
    private var descriptionEditText: EditText? = null
    private var saveEvent: Button? = null
    var locationEditText: EditText? = null
    private var startDateEditText: EditText? = null
    private var locationView: AutocompleteSupportFragment? = null
    private var dateFormat = SimpleDateFormat("dd MMM, yyyy", Locale.getDefault())
    private var myCalendar: Calendar = Calendar.getInstance()
    private var event = Event()
    private var sessionManager: SessionManager? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.event)
        startDateEditText = findViewById(R.id.startDateEditText)
        locationEditText = findViewById(R.id.locationEditText)
        saveEvent = findViewById(R.id.saveEvent)
        descriptionEditText = findViewById(R.id.descriptionEditText)
        nameEditText = findViewById(R.id.nameEditText)
        sessionManager = SessionManager(this)
        event = sessionManager!!.event
        event.startDate?.let {
            myCalendar.time = it
        }
        databaseReference = FirebaseDatabase.getInstance().reference
        val actionBar = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)
        if (!Places.isInitialized()) {
            Places.initialize(applicationContext, "YOUR_API_KEY")
        }
        locationView = supportFragmentManager.findFragmentById(R.id.locationView) as AutocompleteSupportFragment?
        if (locationView != null) {
            locationView!!.setPlaceFields(listOf(Place.Field.ID, Place.Field.NAME))
            locationView!!.setOnPlaceSelectedListener(object : PlaceSelectionListener {
                override fun onPlaceSelected(place: Place) {
                    Log.d(TAG, "Place:" + place.name)
                    if (place.latLng != null) {
                        Log.d(TAG, "Latitude: ${place.latLng?.latitude}")
                        Log.d(TAG, "Longitude: ${place.latLng?.longitude}")
                    }
                    locationEditText?.setText(place.name)
                }

                override fun onError(status: Status) {
                    Log.e(TAG, "Error: $status")
                }
            })
        }
    }

    override fun onStart() {
        super.onStart()
        updateViews()
    }

    //update views
    private fun updateViews() {
        nameEditText!!.setText(event.name)
        descriptionEditText!!.setText(event.description)
        locationEditText!!.setText(event.location)
        event.startDate?.let {
            startDateEditText!!.setText(dateFormat.format(it))
        }

        saveEvent!!.setOnClickListener {
            if (isValid) {
                saveEvent()
            } else {
                Toast.makeText(applicationContext, "Please make sure all fields are filled", Toast.LENGTH_LONG).show()
            }
        }
        val date = OnDateSetListener { _: DatePicker?, year: Int, monthOfYear: Int, dayOfMonth: Int ->
            myCalendar[Calendar.YEAR] = year
            myCalendar[Calendar.MONTH] = monthOfYear
            myCalendar[Calendar.DAY_OF_MONTH] = dayOfMonth
            updateLabel()
        }
        startDateEditText!!.onFocusChangeListener = OnFocusChangeListener { _: View?, hasFocus: Boolean ->
            if (hasFocus) {
                DatePickerDialog(this@EventActivity, date, myCalendar[Calendar.YEAR], myCalendar[Calendar.MONTH],
                        myCalendar[Calendar.DAY_OF_MONTH]).show()
            }
        }
    }

    private val isValid: Boolean
        get() = nameEditText!!.text.toString().isNotEmpty() && descriptionEditText!!.text.toString()
            .isNotEmpty()

    private fun saveEvent() {
        val eventReference: DatabaseReference
        if (event.eventId == null) {
            eventReference = databaseReference!!.child("events").push()
            event.eventId = eventReference.key
        } else {
            eventReference = databaseReference!!.child("events").child(event.eventId!!)
        }
        event.name = nameEditText!!.text.toString()
        event.description = descriptionEditText!!.text.toString()
        event.location = locationEditText!!.text.toString()
        if (FirebaseAuth.getInstance().currentUser != null) {
            val firebaseUser = FirebaseAuth.getInstance().currentUser
            if (firebaseUser != null) {
                event.userId = firebaseUser.uid
            }
        }
        try {
            event.startDate = dateFormat.parse(startDateEditText?.text.toString())
        } catch (e: ParseException) {
            //TODO: do something serious
            //e.printStackTrace();
        }
        eventReference.updateChildren(event.toMap()).addOnCompleteListener { task: Task<Void?> ->
            if (task.isSuccessful) {
                Log.d(TAG, "Success")
                Toast.makeText(this@EventActivity, "Event saved successfully", Toast.LENGTH_SHORT).show()
            } else {
                Log.e(TAG, "erro")
                var errorMessage: String? = null
                val exception = task.exception
                if (exception != null) {
                    errorMessage = exception.message
                }
                Toast.makeText(this@EventActivity, "Error saving Event $errorMessage", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun updateLabel() {
        startDateEditText!!.setText(dateFormat.format(myCalendar.time))
    }

    companion object {
        private const val TAG = "SAPWF"
    }
}